src/Fixer/Whitespace/HeredocIndentationFixer.php

```diff
@@ -0,0 +1,152 @@
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'Heredoc/nowdoc content must be properly indented. Requires PHP >= 7.3.',
+            [
+                new VersionSpecificCodeSample(
+                    <<<'SAMPLE'
+<?php
+    $a = <<<EOD
+abc
+    def
+EOD;
+
+SAMPLE
+                    ,
+                    new VersionSpecification(70300)
+                ),
+                new VersionSpecificCodeSample(
+                    <<<'SAMPLE'
+<?php
+    $a = <<<'EOD'
+abc
+    def
+EOD;
+
+SAMPLE
+                    ,
+                    new VersionSpecification(70300)
+                ),
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return \PHP_VERSION_ID >= 70300 && $tokens->isTokenKindFound(T_START_HEREDOC);
+    }
+
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        for ($index = count($tokens) - 1; 0 <= $index; --$index) {
+            if (!$tokens[$index]->isGivenKind(T_END_HEREDOC)) {
+                continue;
+            }
+
+            $end = $index;
+
+            for (--$index; !$tokens[$index]->isGivenKind(T_START_HEREDOC); --$index);
+
+            $this->fixIndentation($tokens, $index, $end);
+        }
+    }
+
+    /**
+     * @param Tokens $tokens
+     * @param int    $start
+     * @param int    $end
+     */
+    private function fixIndentation(Tokens $tokens, $start, $end)
+    {
+        $indent = $this->getIndentAt($tokens, $start).$this->whitespacesConfig->getIndent();
+
+        Preg::match('/^[ \t]*/', $tokens[$end]->getContent(), $matches);
+        $currentIndent = $matches[0];
+
+        $content = $indent.substr($tokens[$end]->getContent(), strlen($currentIndent));
+        $tokens[$end] = new Token([T_END_HEREDOC, $content]);
+
+        if ($end === $start + 1) {
+            return;
+        }
+
+        for ($index = $end - 1, $last = true; $index > $start; --$index, $last = false) {
+            if (!$tokens[$index]->isGivenKind([T_ENCAPSED_AND_WHITESPACE, T_WHITESPACE])) {
+                continue;
+            }
+
+            $regexEnd = $last && !$currentIndent ? '(?!$)' : '';
+            $content = Preg::replace('/(?<=\v)'.$currentIndent.$regexEnd.'/', $indent, $tokens[$index]->getContent());
+            $tokens[$index] = new Token([$tokens[$index]->getId(), $content]);
+        }
+
+        ++$index;
+
+        if ($tokens[$index]->isGivenKind(T_ENCAPSED_AND_WHITESPACE)) {
+            $content = $indent.substr($tokens[$index]->getContent(), strlen($currentIndent));
+            $tokens[$index] = new Token([T_ENCAPSED_AND_WHITESPACE, $content]);
+        } else {
+            $tokens->insertAt($index, new Token([T_ENCAPSED_AND_WHITESPACE, $indent]));
+        }
+    }
+
+    /**
+     * @param int $index
+     *
+     * @return null|string
```

Comment:
is `null` a real possible value here? the calling code seems to assume a string is always returned (it concats on the returned value)

Reply:
fixed